### PR TITLE
 Add endpoint to fetch DPoS module constants - Closes #1259

### LINKED ISSUE #1259: Add endpoint to fetch DPoS module constants
### Description

With the DPoS module enabled, users should be able to see the constants defined by the module.

### Acceptance Criteria

- The following endpoints are available
- The module constants are cached at init
- Requests are served from the cache
- Integration tests are implemented
- Swagger specs are up-to-date

#### Endpoints
**HTTP**:  `GET /dpos/constants`
**RPC**:    `get.dpos.constants`

#### Request query params
None

#### Response
```
{
    data: {
        factorSelfVotes: 10,
        maxLengthName: 20,
        maxNumberSentVotes: 10,
        maxNumberPendingUnlocks: 20,
        failSafeMissedBlocks: 50,
        failSafeInactiveWindow: 260000,
        punishmentWindow: 780000,
        roundLength: 103,
        bftThreshold: 68,
        minWeightStandby: '100000000000',
        numberActiveDelegates: 101,
        numberStandbyDelegates: 2,
        tokenIDDPoS: '0000000000000000',
    },
    meta: {

    }
}
```

### PR DESCRIPTION
**Pull request by priojeetpriyom**
### What was the problem?
This PR resolves #1259

### How was it solved?

- [x] The following endpoints are available
- [x] The module constants are cached at init - will be cached on first call
- [x] Requests are served from the cache
- [x] Integration tests are implemented
- [x] Swagger specs are up-to-date
- [x] Merge all methods for dpos related endpoints into `services/blockchain-indexer/methods/dataService/controllers/dpos.js`
and `services/blockchain-indexer/methods/dataService/dpos.js`

### CODE REVIEW

File: tests/integration/api_v3/http/dposConstants.test.js [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+const baseUrlV3 = `${baseUrl}/api/v3`;
+
+const {
+	badRequestSchema,
+} = require('../../../schemas/httpGenerics.schema');
+
+const {
+	dposConstantsSchema,
+	dposConstantsMetaSchema,
+} = require('../../../schemas/api_v3/dposConstants.schema');
```
Comment (nagdahimanshu) [inline tests/integration/api_v3/http/dposConstants.test.js:29]:
```suggestion
const {
	badRequestSchema,
} = require('../../../schemas/httpGenerics.schema');

const {
	dposConstantsSchema,
	dposConstantsMetaSchema,
} = require('../../../schemas/api_v3/dposConstants.schema');

const baseUrl = config.SERVICE_ENDPOINT;
const baseUrlV3 = `${baseUrl}/api/v3`;
```

File: tests/integration/api_v3/http/dposConstants.test.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+const endpoint = `${baseUrlV3}/dpos/constants`;
+
+describe('DPoS Constants API', () => {
+	describe(`GET ${endpoint}`, () => {
```
Comment (nagdahimanshu) [inline tests/integration/api_v3/http/dposConstants.test.js:34]: There is no need to define `describe` functions 2 times in this case.

File: tests/integration/api_v3/http/dposConstants.test.js [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+		it('Returns unlocks when requested for existing account by address', async () => {
```
Comment (nagdahimanshu) [inline tests/integration/api_v3/http/dposConstants.test.js:35]:
```suggestion
		it('Returns DPoS module constants', async () => {
```

File: tests/integration/api_v3/http/dposConstants.test.js [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+			const response = await api.get(`${endpoint}`);
+			expect(response.data).toMap(dposConstantsSchema);
+			expect(response.data.minWeightStandby.length).toBeGreaterThanOrEqual(1);
+			expect(response.data.tokenIDDPoS.length).toBeGreaterThanOrEqual(10);
```
Comment (nagdahimanshu) [inline tests/integration/api_v3/http/dposConstants.test.js:39]:
```suggestion
```

File: tests/integration/api_v3/rpc/dposConstants.test.js [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+describe('get.dpos.constants', () => {
+	it('Returns constants from sdk dpos module when requested', async () => {
+		const response = await getDPoSConstants();
+		expect(response).toMap(jsonRpcEnvelopeSchema);
+
+		const { result } = response;
+		expect(result.data).toMap(dposConstantsSchema);
+		expect(result.data.minWeightStandby.length).toBeGreaterThanOrEqual(1);
+		expect(result.data.tokenIDDPoS.length).toBeGreaterThanOrEqual(1);
```
Comment (nagdahimanshu) [inline tests/integration/api_v3/rpc/dposConstants.test.js:41]:
```suggestion
```

File: services/gateway/apis/http-version3/methods/dposConstants.js [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+module.exports = {
+	version: '2.0',
+	swaggerApiPath: '/dpos/constants',
+	rpcMethod: 'get.dpos.constants',
+	tags: ['DPoS'],
+	envelope,
+	params: {},
+	paramsRequired: true,
+	validParamPairings: [],
```
[the remark this answers is not in the document]
Comment (priojeetpriyom) [reply services/gateway/apis/http-version3/methods/dposConstants.js]: Applied.

File: services/gateway/apis/http-version3/methods/dposConstants.js [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+const dposConstantsSource = require('../../../sources/version3/dposConstants');
+const envelope = require('../../../sources/version3/mappings/stdEnvelope');
+const { response, getSwaggerDescription } = require('../../../shared/utils');
+
+module.exports = {
+	version: '2.0',
+	swaggerApiPath: '/dpos/constants',
+	rpcMethod: 'get.dpos.constants',
+	tags: ['DPoS'],
+	envelope,
```
[the remark this answers is not in the document]
Comment (priojeetpriyom) [reply services/gateway/apis/http-version3/methods/dposConstants.js]: some other files also had this. updated all.

File: services/gateway/apis/http-version3/methods/dposConstants.js [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+	params: {},
+	paramsRequired: true,
+	validParamPairings: [],
+	get schema() {
+		const constantsSchema = {};
+		constantsSchema[this.swaggerApiPath] = { get: {} };
+		constantsSchema[this.swaggerApiPath].get.tags = this.tags;
+		constantsSchema[this.swaggerApiPath].get.summary = 'Requests DPoS Constants';
```
[the remark this answers is not in the document]
Comment (priojeetpriyom) [reply services/gateway/apis/http-version3/methods/dposConstants.js]: Applied.

File: tests/integration/api_v3/http/dposConstants.test.js (hunk 1 above)
Comment (priojeetpriyom) [reply tests/integration/api_v3/http/dposConstants.test.js]: dposUnlock was also importing like this. both updated.

File: tests/integration/api_v3/http/dposConstants.test.js (hunk 2 above)
Comment (priojeetpriyom) [reply tests/integration/api_v3/http/dposConstants.test.js]: Applied.

File: tests/integration/api_v3/http/dposConstants.test.js (hunk 3 above)
Comment (priojeetpriyom) [reply tests/integration/api_v3/http/dposConstants.test.js]: Applied.

File: tests/integration/api_v3/http/dposConstants.test.js (hunk 4 above)
Comment (priojeetpriyom) [reply tests/integration/api_v3/http/dposConstants.test.js]: Applied.

File: tests/integration/api_v3/rpc/dposConstants.test.js (hunk 5 above)
Comment (priojeetpriyom) [reply tests/integration/api_v3/rpc/dposConstants.test.js]: Applied.

File: services/gateway/apis/http-version3/methods/dposConstants.js [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+	rpcMethod: 'get.dpos.constants',
+	tags: ['DPoS'],
+	get schema() {
+		const constantsSchema = {};
+		constantsSchema[this.swaggerApiPath] = { get: {} };
+		constantsSchema[this.swaggerApiPath].get.tags = this.tags;
+		constantsSchema[this.swaggerApiPath].get.summary = 'Requests DPoS Module Constants';
+		constantsSchema[this.swaggerApiPath].get.description = getSwaggerDescription({
+			rpcMethod: this.rpcMethod,
+			description: 'Returns a set of constants exposed from DPoS Module',
```
Comment (nagdahimanshu) [inline services/gateway/apis/http-version3/methods/dposConstants.js:33]:
```suggestion
			description: 'Requests all the configured constants for the DPoS module',
```

File: services/gateway/apis/http-version3/methods/dposConstants.js [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+		});
+		constantsSchema[this.swaggerApiPath].get.responses = {
+			200: {
+				description: 'Returns a set of constants exposed from DPoS Module',
```
Comment (nagdahimanshu) [inline services/gateway/apis/http-version3/methods/dposConstants.js:37]:
```suggestion
				  description: 'Returns all the configured constants for the DPoS module',
```

File: services/gateway/apis/http-version3/swagger/definitions/dpos.json [hunk 11]
```diff
@@ -289,5 +289,89 @@
...
+    },
+    "DPoSConstantsWithEnvelope": {
+        "type": "object",
+        "required": [
+            "data",
+            "meta"
+        ],
+        "properties": {
+            "data": {
+                "description": "DPoS Module constants",
```
Comment (nagdahimanshu) [inline services/gateway/apis/http-version3/swagger/definitions/dpos.json:301]:
```suggestion
                "description": "DPoS module constants",
```

File: services/gateway/apis/http-version3/swagger/definitions/dpos.json [hunk 12]
```diff
@@ -289,5 +289,89 @@
...
+                "type": "object",
+                "properties": {
+                    "factorSelfVotes": {
+                        "type": "integer",
+                        "example": 10,
+                        "description": "The factor multiplying the self-votes of a delegate for the delegate weight computation."
```
Comment (nagdahimanshu) [inline services/gateway/apis/http-version3/swagger/definitions/dpos.json:307]:
```suggestion
                        "description": "The factor multiplying the self-votes of a delegate for the delegate weight computation. It is used to set the ceiling for the delegate weight."
```

File: services/gateway/apis/http-version3/swagger/definitions/dpos.json [hunk 13]
```diff
@@ -289,5 +289,89 @@
...
+                    },
+                    "maxLengthName": {
+                        "type": "integer",
+                        "example": 20,
+                        "description": "The maximum allowed name length for delegates."
```
Comment (nagdahimanshu) [inline services/gateway/apis/http-version3/swagger/definitions/dpos.json:312]:
```suggestion
                        "description": "The maximum number of characters allowed for a delegate name."
```

File: services/gateway/apis/http-version3/swagger/definitions/dpos.json [hunk 14]
```diff
@@ -289,5 +289,89 @@
...
+                    },
+                    "maxNumberSentVotes": {
+                        "type": "integer",
+                        "example": 10,
+                        "description": "The maximum size of the sentVotes array of a voter."
```
Comment (nagdahimanshu) [inline services/gateway/apis/http-version3/swagger/definitions/dpos.json:317]:
```suggestion
                        "description": "The maximum number of delegates any voter can vote for."
```

File: services/gateway/apis/http-version3/swagger/definitions/dpos.json [hunk 15]
```diff
@@ -289,5 +289,89 @@
...
+                    },
+                    "maxNumberPendingUnlocks": {
+                        "type": "integer",
+                        "example": 20,
+                        "description": "The maximum size of the pendingUnlocks array of a voter."
```
Comment (nagdahimanshu) [inline services/gateway/apis/http-version3/swagger/definitions/dpos.json:322]:
```suggestion
                        "description": "The maximum number of pending unlocks allowed for a voter."
```

File: services/gateway/apis/http-version3/swagger/definitions/dpos.json [hunk 16]
```diff
@@ -289,5 +289,89 @@
...
+                    },
+                    "failSafeMissedBlocks": {
+                        "type": "integer",
+                        "example": 50,
+                        "description": "The number of consecutive missed blocks used in the fail safe banning mechanism."
```
Comment (nagdahimanshu) [inline services/gateway/apis/http-version3/swagger/definitions/dpos.json:327]:
```suggestion
                        "description": "The number of consecutive missed blocks used in the fail-safe banning mechanism."
```

File: services/gateway/apis/http-version3/swagger/definitions/dpos.json [hunk 17]
```diff
@@ -289,5 +289,89 @@
...
+                    },
+                    "failSafeInactiveWindow": {
+                        "type": "integer",
+                        "example": 260000,
+                        "description": "The length of the inactivity window used in the fail safe banning mechanism."
```
Comment (nagdahimanshu) [inline services/gateway/apis/http-version3/swagger/definitions/dpos.json:332]:
```suggestion
                        "description": "The length of the inactivity window used in the fail-safe banning mechanism."
```

File: services/gateway/apis/http-version3/swagger/definitions/dpos.json [hunk 18]
```diff
@@ -289,5 +289,89 @@
...
+                    },
+                    "punishmentWindow": {
+                        "type": "integer",
+                        "example": 780000,
+                        "description": "The punishment time for punished delegates."
+                    },
+                    "roundLength": {
+                        "type": "integer",
+                        "example": 10,
+                        "description": "The round length i.e the sum of active and standby delegates"
```
Comment (nagdahimanshu) [inline services/gateway/apis/http-version3/swagger/definitions/dpos.json:342]:
```suggestion
                        "description": "The number of blocks in a round and is the sum total of the active and the standby delegates."
```

File: services/gateway/apis/http-version3/swagger/definitions/dpos.json [hunk 19]
```diff
@@ -289,5 +289,89 @@
...
+                    },
+                    "bftThreshold": {
+                        "type": "integer",
+                        "example": 68,
+                        "description": "The precommit and certificate thresholds used by the BFT module."
```
Comment (nagdahimanshu) [inline services/gateway/apis/http-version3/swagger/definitions/dpos.json:347]:
```suggestion
                        "description": "The pre-commit and certificate thresholds used by the BFT module."
```

File: services/gateway/apis/http-version3/swagger/definitions/dpos.json [hunk 20]
```diff
@@ -289,5 +289,89 @@
...
+                    },
+                    "minWeightStandby": {
+                        "type": "string",
+                        "example": "100000000000",
+                        "description": "The minimum delegate weight required to be eligible as a standby delegate."
+                    },
+                    "numberActiveDelegates": {
+                        "type": "integer",
+                        "example": 101,
+                        "description": "The number of active delegates."
```
Comment (nagdahimanshu) [inline services/gateway/apis/http-version3/swagger/definitions/dpos.json:357]:
```suggestion
                        "description": "The number of active delegates in a round."
```

File: services/gateway/apis/http-version3/swagger/definitions/dpos.json [hunk 21]
```diff
@@ -289,5 +289,89 @@
...
+                    },
+                    "numberStandbyDelegates": {
+                        "type": "integer",
+                        "example": 2,
+                        "description": "The number of standby delegates."
```
Comment (nagdahimanshu) [inline services/gateway/apis/http-version3/swagger/definitions/dpos.json:362]:
```suggestion
                        "description": "The number of standby delegates in a round."
```

File: tests/integration/api_v3/http/dposConstants.test.js [hunk 22]
```diff
@@ -0,0 +1,46 @@
...
+	dposConstantsSchema,
+	dposConstantsMetaSchema,
+} = require('../../../schemas/api_v3/dposConstants.schema');
+
+const baseUrl = config.SERVICE_ENDPOINT;
+const baseUrlV3 = `${baseUrl}/api/v3`;
+const endpoint = `${baseUrlV3}/dpos/constants`;
+
+describe('DPoS Constants API', () => {
+	it('Returns DPoS module constants when called correctly', async () => {
```
Comment (nagdahimanshu) [inline tests/integration/api_v3/http/dposConstants.test.js:33]:
```suggestion
	it('returns DPoS module constants when called correctly', async () => {
```

File: tests/integration/api_v3/rpc/dposConstants.test.js [hunk 23]
```diff
@@ -0,0 +1,52 @@
...
+	dposConstantsSchema,
+	dposConstantsMetaSchema,
+} = require('../../../schemas/api_v3/dposConstants.schema');
+
+const wsRpcUrl = `${config.SERVICE_ENDPOINT}/rpc-v3`;
+
+const getDPoSConstants = async (params) => request(wsRpcUrl, 'get.dpos.constants', params);
+
+describe('get.dpos.constants', () => {
+	it('Returns constants from sdk dpos module when requested', async () => {
```
Comment (nagdahimanshu) [inline tests/integration/api_v3/rpc/dposConstants.test.js:34]:
```suggestion
	it('returns constants from sdk dpos module when requested', async () => {
```

**Review by nagdahimanshu — CHANGES_REQUESTED** [review]

**Review by nagdahimanshu — APPROVED** [review]
Looks good to me 👍

**Pull request merged**